Add support for connected devices

matthewellis opened the pull request:
Moved and improved from Fastlane core repo. 
https://github.com/fastlane/fastlane_core/pull/120

This change adds support for usb connected tvOS and iOS devices

matthewellis committed:
Add support for connected devices

mfurtak commented:
This seems great - thanks for the contribution! :+1: 

It would also be great to see some tests for this functionality. Do you think you can start with some canned output from `system_profiler SPUSBDataType` and `xcrun simctl list devices` in `spec/fixtures` and make some assertions from there?

KrauseFx commented on fastlane_core/lib/fastlane_core/device_manager.rb:
```diff
@@ -40,20 +40,64 @@ def all
+        end
+
+        usb_devices_output = ''
+        Open3.popen3("system_profiler SPUSBDataType") do |stdin, stdout, stderr, wait_thr|
```
:heart:

KrauseFx commented on fastlane_core/lib/fastlane_core/device_manager.rb:
```diff
@@ -40,20 +40,64 @@ def all
+          line = usb_devices_output_lines[line_index]
+          is_supported_device = !line.nil? && device_types.any? { |device_type| line.include?(device_type) }
+          next unless is_supported_device
+          usb_info_lines = 10
```
This seems hacky. Instead you could use
```
system_profiler SPUSBDataType -xml
```
to get the output in XML so we can parse it. Please don't add `nokogiri` as a dependency, but instead use the `plist` gem :+1:

mfurtak replied:
Oooh, very nice!

matthewellis replied:
Perfect, i wasn't too happy with my solution.

KrauseFx commented:
Thanks @matthewellis

KrauseFx commented:
Thanks @matthewellis, I added a comment regarding how we read the output of the USB devices. There seems to be no JSON API, however we can read this data using XML, so that's better than manually parsing the text :+1:

matthewellis commented:
Updated my pull request to use xml output of system_profiler and add refactored simulator_spec into device_manager_spec to add tests for connected devices and a test for all simulator types. I also moved the output of simctl to a file in the fixtures folder. 

Let me know if there is anything else i need to tidy up. :rocket:

mfurtak commented on fastlane_core/spec/device_manager_spec.rb:
```diff
@@ -0,0 +1,264 @@
+    before do
+      @valid_simulators = File.read('./spec/fixtures/DeviceManagerSimctlOutput')
+      FastlaneCore::Simulator.clear_cache
+    end
```
suggestion: change this to a `before(:all)` block and move all fixture file loading in. This way we can just read the canned responses once, instead of before each individual test.

matthewellis replied:
Good idea, will update tonight! :+1:

mfurtak commented on fastlane_core/lib/fastlane_core/device_manager.rb:
```diff
@@ -40,20 +41,59 @@ def all
+          device_types = ["AppleTV"]
+        elsif requested_os_type == "iOS"
+          device_types = ["iPhone", "iPad"]
+        end
```
Suggestion: This can be tightened up a bit like:

```ruby
device_types = if requested_os_type == "tvOS"
                 ["AppleTV"]
               elsif requested_os_type == "iOS"
                 ["iPhone", "iPad"]
               else 
                 []
               end
```

matthewellis commented:
Are we all good to merge? :smile:

mfurtak commented on fastlane_core/lib/fastlane_core/device_manager.rb:
```diff
@@ -40,20 +41,60 @@ def all
+        device_types = if requested_os_type == "tvOS"
+                         ["AppleTV"]
+                       elsif requested_os_type == "iOS"
+                         ["iPhone", "iPad"]
```
Sorry to have missed it earlier... let's include "iPod" as well! :smile:

matthewellis commented:
Made the changes to include iPod's :+1:

mfurtak commented:
Great, I will have a look tomorrow!

mfurtak commented on fastlane_core/lib/fastlane_core/device_manager.rb:
```diff
@@ -126,18 +169,38 @@ def reset
   end
 
-  class SimulatorWatch < Simulator
+  class SimulatorWatch < DeviceManager
```
Are these classes gaining anything by being subclasses of `DeviceManager`? It seems like all interactions between them are done by direct class method calls. (`DeviceManager.simulators`, etc)

mfurtak replied:
I did a little experimenting and answered my own question. Given that `DeviceManager`'s methods become available on these classes, does it make sense to be able to say

```ruby
SimulatorWatch.connected_devices('tvOS')
```

matthewellis replied:
Good point, i think this had been due to prior experimentation when developing this but not removed when cleaning up. Will amend my commit and separate the classes.

matthewellis commented:
Fixed subclass, could you have a look when you get a chance?

mfurtak commented:
Looking good to me. @KrauseFx what are your thoughts?

KrauseFx commented:
Hey @mfurtak, sorry I somehow missed that notification, thanks for merging :+1: I'll take a look at the source code now

KrauseFx commented:
This is great, thank you so much for working on this 👍
